content/microservices/nodejs/tabs/cdk.md

```diff
@@ -201,6 +201,149 @@ cdk deploy
+    - If we let the orchestrator handle the scaling in and out for the service, we can focus on continuous improvement, and less on operational heavy lifting. In order to get autoscaling setup, one first needs to know what metric to use as the decision to autoscale. Some example metrics for scaling are CPU utilization, memory utilization, and queue depth.
+
+#### Setup Autoscaling in the code
+
+- Using the editor of your choice, open 'environment/ecsdemo-nodejs/cdk/app.py' in the cdk directory.
+
+- Search for `Enable Service Autoscaling` to find the code that will enable autoscaling for the service.
+
+- Remove the comments (#) from the code for self.autoscale and below, once you remove them, it should look like the following:
+
+```python
+# Enable Service Autoscaling
+self.autoscale = self.fargate_service.auto_scale_task_count(
+    min_capacity=1,
+    max_capacity=10
+)
+
+self.autoscale.scale_on_cpu_utilization(
+    "CPUAutoscaling",
+    target_utilization_percent=50,
+    scale_in_cooldown=core.Duration.seconds(30),
+    scale_out_cooldown=core.Duration.seconds(30)
+)
+
+self.stressToolEc2 = self.nodejsTempEc2StressTool()
+```
+
+### Install the session Manager plugin into cloud9 so we can connect to the ec2 instance
+
+```shell
+curl "https://s3.amazonaws.com/session-manager-downloads/plugin/latest/linux_64bit/session-manager-plugin.rpm" -o "session-manager-plugin.rpm"
+sudo yum install -y session-manager-plugin.rpm
+```
+
+#### Verify the plugin is working
+
+Run the following to command to verify the installation
+```shell
+session-manager-plugin
+```
+
+you should get something like this output:
+![verify-ssm-agent-plugin](/images/cdk-ssm-plugin-installation-output.png)
+
+{{% /expand %}}
+
+#### Code Review
+
+- To start modeling our autoscaling logic, we first set what our upper and lower bounds are. This ensures that we will always be at a minimum of 1 task, and a maximum of 10 tasks.
+
+```python
+# Enable Service Autoscaling
+self.autoscale = self.fargate_service.auto_scale_task_count(
+    min_capacity=1,
+    max_capacity=10
+)
+```
+
+- When the ECS service is deployed, Cloudwatch metrics such as CPU utilization are enabled by default. We are going to take advantage of that metric and use it as our scaling target. In this method, we are setting what our target cpu utilization percent is, and how long in between scale activities we want to wait before it adds/removes another task.
+
+```python
+self.autoscale.scale_on_cpu_utilization(
+    "CPUAutoscaling",
+    target_utilization_percent=50,
+    scale_in_cooldown=core.Duration.seconds(30),
+    scale_out_cooldown=core.Duration.seconds(30)
+)
+```
+
+- To stress the nodejs fargate service we need to have the ability to reach out the ECS nodejs fargate service end point. Since the nodejs service is in a private subnet, we will deploy an EC2 into the same VPC and use that EC2 instance to stress the nodejs service and see the auto-scale in action.
```

Comment:
In order to introduce load to the Nodejs Fargate service, we need to have the ability to reach its service endpoint. Since the service is in a private subnet, we will deploy an EC2 instance into the same VPC, and use that EC2 instance to put stress on the service. The additional stress will trigger an auto scaling event.

Reply:
I'll do the change.